jerry-core/ecma/builtin-objects/ecma-builtin-generator-prototype.c

```diff
@@ -109,27 +101,98 @@ ecma_builtin_generator_prototype_object_do (ecma_value_t this_arg, /**< this arg
     return ecma_create_iter_result_object (ECMA_VALUE_UNDEFINED, ECMA_VALUE_TRUE);
   }
 
-  if (resume_mode == ECMA_GENERATOR_RETURN)
-  {
-    executable_object_p->frame_ctx.byte_code_p = ecma_builtin_generator_prototype_return;
-  }
-  else if (resume_mode == ECMA_GENERATOR_THROW)
+  arg = ecma_copy_value (arg);
+
+  while (true)
   {
-    executable_object_p->frame_ctx.byte_code_p = ecma_builtin_generator_prototype_throw;
-  }
+    if (executable_object_p->extended_object.u.class_prop.extra_info & ECMA_GENERATOR_ITERATE_AND_YIELD)
+    {
+      ecma_value_t iterator = executable_object_p->extended_object.u.class_prop.u.value;
 
-  ecma_value_t value = opfunc_resume_executable_object (executable_object_p, arg);
+      bool done = false;
+      ecma_value_t result = ecma_op_iterator_do (resume_mode, iterator, arg, &done);
+      ecma_free_value (arg);
 
-  if (JERRY_UNLIKELY (ECMA_IS_VALUE_ERROR (value)))
-  {
-    return value;
-  }
+      if (ECMA_IS_VALUE_ERROR (result))
+      {
+        arg = result;
+      }
+      else if (done)
+      {
+        arg = ecma_op_iterator_value (result);
+        ecma_free_value (result);
+        if (resume_mode == ECMA_ITERATOR_THROW)
+        {
+          /* This part is changed in the newest ECMAScript standard.
+           * It was ECMA_ITERATOR_RETURN in ES2015, but I think it was a typo. */
+          resume_mode = ECMA_ITERATOR_NEXT;
+        }
+      }
+      else
+      {
+        return result;
+      }
 
-  bool done = (executable_object_p->extended_object.u.class_prop.extra_info & ECMA_EXECUTABLE_OBJECT_COMPLETED);
-  ecma_value_t result = ecma_create_iter_result_object (value, ecma_make_boolean_value (done));
+      executable_object_p->extended_object.u.class_prop.extra_info &= (uint16_t) ~ECMA_GENERATOR_ITERATE_AND_YIELD;
 
-  ecma_fast_free_value (value);
-  return result;
+      if (ECMA_IS_VALUE_ERROR (arg))
+      {
+        arg = JERRY_CONTEXT (error_value);
+        JERRY_CONTEXT (status_flags) &= (uint32_t) ~ECMA_STATUS_EXCEPTION;
+        resume_mode = ECMA_ITERATOR_THROW;
+      }
+    }
+
+    if (resume_mode == ECMA_ITERATOR_RETURN)
+    {
+      executable_object_p->frame_ctx.byte_code_p = ecma_builtin_generator_prototype_return;
+    }
+    else if (resume_mode == ECMA_ITERATOR_THROW)
+    {
+      executable_object_p->frame_ctx.byte_code_p = ecma_builtin_generator_prototype_throw;
+    }
+
+    ecma_value_t value = opfunc_resume_executable_object (executable_object_p, arg);
+
+    if (JERRY_UNLIKELY (ECMA_IS_VALUE_ERROR (value)))
```

Comment:
The `ECMA_IS_VALUE_ERROR` already contains a `JERRY_UNLIKELY` call.

Reply:
Done